lib/fswatcher/fswatcher.go

```diff
@@ -0,0 +1,496 @@
+			oldEvents[path] = event
+		}
+	}
+	for path, event := range dir.events {
+		// 2 * results in mean event age of notifyDelay (assuming randomly
+		// occurring events). Reoccuring and remove (for efficient renames)
+		// events are delayed until notifyTimeout.
+		if (event.eventType != remove && 2*currTime.Sub(event.lastModTime) > watcher.notifyDelay) ||
+			currTime.Sub(event.firstModTime) > watcher.notifyTimeout {
+			oldEvents[path] = event
+			delete(dir.events, path)
+		}
+	}
+	if dir.parentDir != nil && dir.childCount() == 0 {
+		dir.parentDir.removeEmptyDir(dir.path)
+	}
+	return oldEvents
+}
+
+func (watcher *fsWatcher) updateInProgressSet(event events.Event) {
+	if event.Type == events.ItemStarted {
+		path := event.Data.(map[string]string)["item"]
+		watcher.inProgress[path] = struct{}{}
+	} else if event.Type == events.ItemFinished {
+		path := event.Data.(map[string]interface{})["item"].(string)
+		delete(watcher.inProgress, path)
+	}
+}
+
+func (watcher *fsWatcher) pathInProgress(path string) bool {
+	_, exists := watcher.inProgress[path]
+	return exists
+}
+
+func (watcher *fsWatcher) UpdateIgnores(ignores *ignore.Matcher) {
+	l.Debugln(watcher, "Ignore patterns update")
+	watcher.ignoresUpdate <- ignores
+}
+
+func (watcher *fsWatcher) String() string {
+	return fmt.Sprintf("fswatcher/%s:", watcher.description)
+}
+
+func (dir *eventDir) eventCount() int {
+	count := len(dir.events)
+	for _, dir := range dir.dirs {
+		count += dir.eventCount()
+	}
+	return count
+}
+
+func (dir *eventDir) childCount() int {
+	return len(dir.events) + len(dir.dirs)
+}
+
+func (dir *eventDir) removeEmptyDir(path string) {
+	delete(dir.dirs, path)
+	if dir.parentDir != nil && dir.childCount() == 0 {
+		dir.parentDir.removeEmptyDir(dir.path)
+	}
+}
+
+func watchesLimitTooLowError(folder string) error {
+	return errors.New("Failed to install inotify handler for " +
+		folder +
+		". Please increase inotify limits," +
+		" see http://bit.ly/1PxkdUC for more information.")
+}
+
+func (dir eventDir) getFirstModTime() time.Time {
+	if dir.childCount() == 0 {
+		panic("getFirstModTime must not be used on empty eventDir")
+	}
+	firstModTime := time.Now()
+	for _, childDir := range dir.dirs {
+		dirTime := childDir.getFirstModTime()
+		if dirTime.Before(firstModTime) {
+			firstModTime = dirTime
+		}
+	}
+	for _, event := range dir.events {
+		if event.firstModTime.Before(firstModTime) {
+			firstModTime = event.firstModTime
+		}
+	}
+	return firstModTime
+}
+
+func (dir eventDir) getEventType() fsEventType {
+	if dir.childCount() == 0 {
+		panic("getEventType must not be used on empty eventDir")
+	}
+	var eventType fsEventType
+	for _, childDir := range dir.dirs {
+		eventType |= childDir.getEventType()
+		if eventType == mixed {
+			return mixed
+		}
+	}
+	for _, event := range dir.events {
+		eventType |= event.eventType
+		if eventType == mixed {
+			return mixed
+		}
+	}
+	return eventType
+}
+
+func notifyTimeout(eventDelayS int) time.Duration {
```

Comment:
Done, but I fear it has not gotten much clearer, as these numbers are truly magic (as in I cannot give a simple derivation, just launch into a lengthy description why it felt like they made sense).